packages/lesswrong/server/autocompleteEndpoint.ts

```diff
@@ -0,0 +1,428 @@
+  return `${getPostBodyFormatted(parentPost, revisionsMap, authorsMap)}
+---
+${parentComments.map((comment) => getCommentBodyFormatted(comment, revisionsMap, authorsMap, parentPost)).join("\n---\n")}
+---
+${getCommentBodyFormatted(comment, revisionsMap, authorsMap, parentPost)}
+---
+Comment on ${parentPost.title}
+${currentUser.displayName} 1h ${Math.floor(20 + (Math.random() * 75))} ${Math.floor((Math.random() - 0.5) * 100)}
+${prefix}`.trim();
+}
+
+async function constructMessageHistory(
+  prefix: string,
+  commentIds: string[],
+  postIds: string[],
+  currentUser: any,
+  replyingCommentId?: string,
+  postId?: string
+): Promise<PromptCachingBetaMessageParam[]> {
+  const messages: PromptCachingBetaMessageParam[] = [];
+
+  // Make the fetches parallel to save time
+  const [revisions, posts, comments] = await Promise.all([
+    Revisions.find({ documentId: { $in: [...postIds, ...commentIds] }, fieldName: "contents", version: "1.0.0" }).fetch(),
+    Posts.find({ _id: { $in: postIds } }).fetch(),
+    Comments.find({ _id: { $in: commentIds } }).fetch(),
+  ]);
+
+  const postsOfComments = await Posts.find({ _id: { $in: comments.map((comment) => comment.postId) } }).fetch();
+
+  const authors = await Users.find({ _id: { $in: [...posts.map((post) => post.userId), ...comments.map((comment) => comment.userId)] } }).fetch();
+
+  const authorsMap = new Map(authors.map((author) => [author._id, author]));
+  const revisionsMap = new Map(revisions.map((revision) => [revision.documentId!, revision]));
+
+  // eslint-disable-next-line no-console
+  console.log(`Converting ${posts.length} posts and ${comments.length} comments to messages`);
+  // eslint-disable-next-line no-console
+  console.time("constructPostMessageHistory");
+
+  // Add fetched posts and comments to message history
+  for (const post of posts) {
+    messages.push({
+      role: "user",
+      content: [{ type: "text", text: `<cmd>cat lw/${post._id}.txt</cmd>` }],
+    });
+
+    messages.push({
+      role: "assistant",
+      content: [
+        {
+          type: "text",
+          text: getPostBodyFormatted(post, revisionsMap, authorsMap),
+        },
+      ],
+    });
+  }
+
+  // eslint-disable-next-line no-console
+  console.timeEnd("constructPostMessageHistory");
+
+  for (const comment of comments) {
+    const parentPost = postsOfComments.find((post) => post._id === comment.postId);
+    messages.push({
+      role: "user",
+      content: [{ type: "text", text: `<cmd>cat lw/${comment._id}.txt</cmd>` }],
+    });
+
+    messages.push({
+      role: "assistant",
+      content: [{ type: "text", text: getCommentBodyFormatted(comment, revisionsMap, authorsMap, parentPost!) }],
+    });
+  }
+
+  // Add final user message with prefix
+  messages.push({
+    role: "user",
+    content: [{ type: "text", text: `<cmd>cat lw/hsqKp56whpPEQns3Z.txt</cmd>`, cache_control: {type: "ephemeral"}}]
+  });
+
+  if (replyingCommentId) {
```

Comment:
I would probably shoot this off in a promise at the very top rather than waiting for all of the above to complete, since we don't actually need anything from the block above for this to run.  (I do see that we're using the `authorMap` and `revisionMap`, but I think both of those go away once you start fetching posts/comments using `fetchFragment` or `runQuery`, and in any case they're only used to get any individual comment's revision/author, so you could pass in new instances of them if you didn't want to refactor all of that code.)